test/systemtest/earlgrey/test_sim_verilator.py

```diff
@@ -0,0 +1,164 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+import logging
+import os
+import re
+import time
+from pathlib import Path
+from typing import Optional
+
+import pytest
+
+from .. import config, utils
+
+log = logging.getLogger(__name__)
+
+
+class VerilatorSimEarlgrey:
+    """ A verilator simulation of the Earl Grey toplevel """
+    UART0_SPEED = 9600  # see device/lib/arch/device_sim_verilator.c
+
+    def __init__(self, sim_path: Path, rom_elf_path: Path, work_dir: Path):
+        assert sim_path.is_file()
+        self._sim_path = sim_path
+
+        assert rom_elf_path.is_file()
+        self._rom_elf_path = rom_elf_path
+
+        assert work_dir.is_dir()
+        self._work_dir = work_dir
+
+        self._log = logging.getLogger(__name__)
+
+    def run(self, flash_elf: Optional[Path]):
+        """
+        Run the simulation
+        """
+
+        cmd_sim = [self._sim_path, '--meminit=rom,' + str(self._rom_elf_path)]
+
+        if flash_elf is not None:
+            assert flash_elf.is_file()
+            cmd_sim.append('--meminit=flash,' + str(flash_elf))
```

Comment:
I wonder if continuing to use `spiflash` here is useful? As the README says:

> System tests are end-to-end tests for the whole OpenTitan system. They operate
on build outputs (typically in `build-bin`) and can be used to check that the
full system, consisting of software, hardware, and tooling, works as expected.

`spiflash` is part of that full system so including it as part of these tests seems like a good idea.

Reply:
Yes, using/testing spiflash is useful, and I'm using it for the upcoming FPGA tests. For the Verilator tests, we don't always want to use it, as it is significantly slower than the memload approach. I do have a commit ready to add a single test to use spiflash with the Verilator simulation to test the interaction. However, I'd prefer to do that in a separate PR, as I'm currently debugging some spiflash/bootstrap interaction with the FPGA build which gives false positives on test runs, and I'd like to understand first that the test actually work reliably.

Reply:
Update: the current PR contains both code to run spiflash on the nexysvideo, and on the verilator simulation.